### PR TITLE
Add numpy ndarray serialisation to arrow

### PR DESCRIPTION
**Pull request by donatasm**
Add multidimensional `numpy` array support in layer datasets.

Arrays must be wrapped with the `layer.Arrays` class to provide the conversions between the `numpy.ndarray ` and `arrow` types. Numpy arrays are serialised to `pyarrow.binary` type, using [numpy.save](https://numpy.org/doc/stable/reference/generated/numpy.save.html) and deserialised using [numpy.load](https://numpy.org/doc/stable/reference/generated/numpy.load.html).

```python
import pandas as pd
import numpy as np

import layer
from layer.decorators import dataset

@dataset("ndarrays_demo")
def build():
    return pd.DataFrame(
        {
            "array": layer.Arrays(
                (
                    np.array([[1, 2, 3], [4, 5, 6]], dtype=np.int64),
                    np.array([[7, 8, 9], [10, 11, 12]], dtype=np.int64),
                    np.array([[13, 14, 15], [16, 17, 18]], dtype=np.int64),
                )
            )
        },
        copy=False,
    )
```

### CODE REVIEW

File: layer/pandas_extensions.py [hunk 1]
```diff
@@ -170,8 +170,137 @@ def _image_bytes(image: Image) -> bytes:
...
+        total_bytes = 0
+        for arr in self._arrays:
+            total_bytes += arr.nbytes
+        return total_bytes
+
+
+def _array_bytes(arr: np.ndarray) -> bytes:  # type: ignore
+    with io.BytesIO() as buf:
+        np.save(buf, arr, allow_pickle=False, fix_imports=False)  # type: ignore
+        return buf.getvalue()
```
Comment (cihati) [inline layer/pandas_extensions.py:296]: nit: why is this not a method of `class Arrays`. same for `_ArrayType` and `_load_array`

**Review by cihati — APPROVED** [review]
Presumably, we'll update the docs for the newly exposed `Images` and `Arrays` in a follow-up PR?

**Comment by donatasm** [comment]
> Presumably, we'll update the docs for the newly exposed `Images` and `Arrays` in a follow-up PR?

Yes, the docs are incoming.

File: layer/pandas_extensions.py (hunk 1 above)
Comment (donatasm) [reply layer/pandas_extensions.py:296]: > nit: why is this not a method of `class Arrays`. same for `_ArrayType` and `_load_array`

As `self` is not being accessed, I prefer keeping them as module functions.

**Pull request merged**